pinot-core/src/main/java/com/linkedin/pinot/core/query/aggregation/groupby/AggregationGroupByTrimmingService.java

```diff
@@ -93,46 +85,75 @@ public AggregationGroupByTrimmingService(@Nonnull AggregationFunction[] aggregat
       return trimmedResults;
     }
 
+    // Trim the result only if the size of intermediateResultsMap is larger than the threshold
     if (intermediateResultsMap.size() > _trimThreshold) {
-      // Need to trim.
-
-      // Construct the priority queues.
-      @SuppressWarnings("unchecked")
-      PriorityQueue<GroupKeyResultPair>[] priorityQueues = new PriorityQueue[_numAggregationFunctions];
-      Object[] sampleResults = intermediateResultsMap.values().iterator().next();
+      // Construct the sorters
+      // For comparable intermediate result, use PriorityQueue as the sorter
+      // For incomparable intermediate result, use TreeMapBasedSorter as the sorter
+      Object[] sorters = new Object[_numAggregationFunctions];
       for (int i = 0; i < _numAggregationFunctions; i++) {
-        if (sampleResults[i] instanceof Comparable) {
-          priorityQueues[i] = new PriorityQueue<>(_trimSize + 1, getGroupKeyResultPairComparator(_minOrders[i]));
+        if (_aggregationFunctions[i].isIntermediateResultComparable()) {
+          if (_minOrders[i]) {
+            sorters[i] = new PriorityQueue<GroupKeyResultPair>(_trimSize, Collections.reverseOrder());
+          } else {
+            sorters[i] = new PriorityQueue<GroupKeyResultPair>(_trimSize, new ComparableComparator());
+          }
+        } else {
+          // NOTE: reverse the comparator so that keys are ordered in descending order
+          if (_minOrders[i]) {
+            sorters[i] = new TreeMapBasedSorter(_trimSize, new ComparableComparator());
+          } else {
+            sorters[i] = new TreeMapBasedSorter(_trimSize, Collections.reverseOrder());
+          }
         }
       }
 
-      // Fill results into the priority queues.
+      // Fill results into the sorting structure
       for (Map.Entry<String, Object[]> entry : intermediateResultsMap.entrySet()) {
         String groupKey = entry.getKey();
         Object[] intermediateResults = entry.getValue();
         for (int i = 0; i < _numAggregationFunctions; i++) {
-          PriorityQueue<GroupKeyResultPair> priorityQueue = priorityQueues[i];
-          if (priorityQueue == null) {
-            trimmedResults.get(i).put(groupKey, intermediateResults[i]);
+          Object intermediateResult = intermediateResults[i];
+          if (_aggregationFunctions[i].isIntermediateResultComparable()) {
+            GroupKeyResultPair value = new GroupKeyResultPair(groupKey, (Comparable) intermediateResult);
+            addToPriorityQueue((PriorityQueue<GroupKeyResultPair>) sorters[i], value, _trimSize);
           } else {
-            GroupKeyResultPair newValue = new GroupKeyResultPair(groupKey, (Comparable) intermediateResults[i]);
-            addToPriorityQueue(priorityQueue, newValue, _trimSize);
+            Comparable finalResult = _aggregationFunctions[i].extractFinalResult(intermediateResult);
```

Comment:
You we make intermediate result comparable, then for each comparison (log2m times per insertion or removal from heap), we need to calculate final result once. Implementing this way to calculate final result only once and cache all of them.